packages/owlbot-bootstrapper/common-container/api-field-fetcher.ts

```diff
@@ -0,0 +1,197 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {Octokit} from '@octokit/rest';
+import {Storage} from '@google-cloud/storage';
+import {ApiFields, DriftApi, GHDir, GHFile} from './interfaces';
+import yaml from 'js-yaml';
+import {getWellKnownFileContents, INTER_CONTAINER_VARS_FILE} from './utils';
+import * as fs from 'fs';
+/**
+ * Fetches API-related information
+ *
+ * @param apiId the unique API ID
+ * @param storageClient a new instance of Storage
+ * @param octokit a new instance of Octokit
+ */
+export class ApiFieldFetcher {
```

Comment:
If you're worried about having these extra public methods, you can use `interface` definitions to "hide" them.

For example, you could make `ApiFieldFetcher` be a general interface:

```ts
export interface ApiFieldFetcher {
  async getAndSaveApiInformation(directoryPath: string) => Promise<void>;
}

export class DriftApiFieldFetcher implements ApiFieldFetcher {
  // implementation you have
}
```

Then callers will only interact with a `ApiFieldFetcher` implementation (they don't care what the implementation is) and can only "see" the desired exposed API.

Also, by writing out this interface, we can see that it seems weird that the "fetcher"'s only method involves writing a file. I would expect a fetcher to return a value to be processed later. In that way, I might write the interface to be:

```ts
export interface ApiFieldFetcher {
  loadApiFields(): Promise<ApiFields>;
}
```

The code to write the file to disk would be handled elsewhere (separate the responsibilities into smaller chunks).

Reply:
Some folks might view this approach as overkill and more Java-ish which it is, but declaring an interface is pretty inexpensive in my opinion.

Reply:
Spoke about this earlier; if we decide to add in a second class, we can implement this.